### PR TITLE
Reorganize Layer API

### PR DESCRIPTION
**Pull request by benson31**
This is mostly a reorg of the API, with the removal of some functions that are not used or not needed. Overlaps a smidge with #2334, so I recommend merging that first and then I'll de-conflict.

### CODE REVIEW

File: include/lbann/layers/layer.hpp [hunk 1]
```diff
@@ -467,61 +451,88 @@ class Layer
...
-  /** @brief Set the model that manages this layer. */
-  inline void set_model(model* m) { m_model = m; }
+  // (trb 10/03/2023): used
+  El::Int get_num_spliting_groups() const { return m_num_spliting_groups; }
 
-  virtual El::Matrix<El::Int>* get_sample_indices_per_mb() { return nullptr; };
+  // (trb 10/03/2023): USED BUT THIS IS BAD BECAUSE m_mygrid IS NEVER SET!
+  std::shared_ptr<El::Grid> get_mygrid() const
+  {
+    LBANN_ERROR("This funciton should not be used.");
```
Comment (tbennun) [inline include/lbann/layers/layer.hpp:504]:
```suggestion
    LBANN_ERROR("This function should not be used.");
```

**Comment by benson31** [comment]
This "passes" CI (🙄).

**Comment by tbennun** [comment]
Done merging #2334

**Comment by benson31** [comment]
Rebased on `develop` after merging #2334.

**Review by tbennun — APPROVED** [review]

**Pull request merged**